File: ui/provenance_model.js
/**
 * Provenance Model for UI Request Adapter (WP-2)
 *
 * Tracks immutable request, revision, provider, and fallback metadata
 * without exposing secrets per Section 14.6.
 */

export const ProvenanceSource = {
  LOCAL: 'local',
  NIM: 'nim',
  NIM_PARTIAL_FALLBACK: 'nim_partial_fallback'
};

export class ProvenanceModel {
  constructor({
    source = ProvenanceSource.LOCAL,
    provider = 'nvidia_nim',
    model_id = null,
    base_url_label = null,
    generated_at = new Date().toISOString(),
    request_id = null,
    source_revision = null,
    fallback_scene_ids = [],
    validation_warnings = []
  } = {}) {
    this.source = source;
    this.provider = provider;
    this.model_id = model_id;
    this.base_url_label = base_url_label;
    this.generated_at = generated_at;
    this.request_id = request_id;
    this.source_revision = source_revision;
    this.fallback_scene_ids = Array.isArray(fallback_scene_ids) ? [...fallback_scene_ids] : [];
    this.validation_warnings = Array.isArray(validation_warnings) ? [...validation_warnings] : [];
  }

  toDict() {
    return {
      source: this.source,
      provider: this.provider,
      model_id: this.model_id,
      base_url_label: this.base_url_label,
      generated_at: this.generated_at,
      request_id: this.request_id,
      source_revision: this.source_revision,
      fallback_scene_ids: [...this.fallback_scene_ids],
      validation_warnings: [...this.validation_warnings]
    };
  }

  static createLocal(requestId = null, sourceRevision = null) {
    return new ProvenanceModel({
      source: ProvenanceSource.LOCAL,
      provider: 'local_deterministic',
      request_id: requestId,
      source_revision: sourceRevision
    });
  }

  static fromNimResponse(responseData = {}, requestId = null, sourceRevision = null) {
    const rawSource = responseData?.provenance?.source || responseData?.source;
    let source = ProvenanceSource.NIM;
    if (rawSource === ProvenanceSource.LOCAL || rawSource === ProvenanceSource.NIM_PARTIAL_FALLBACK) {
      source = rawSource;
    }

    return new ProvenanceModel({
      source,
      provider: responseData.provider || 'nvidia_nim',
      model_id: responseData.model_id || null,
      base_url_label: responseData.base_url_label || null,
      generated_at: responseData.generated_at || new Date().toISOString(),
      request_id: responseData.request_id || requestId,
      source_revision: responseData.source_revision || sourceRevision,
      fallback_scene_ids: responseData.fallback_scene_ids || [],
      validation_warnings: responseData.validation_warnings || []
    });
  }
}

File: src/provenance_model.py
"""
Provenance Model for UI Request Adapter (WP-2)

Defines immutable provenance tracking structures for UI requests, responses,
and model metadata without secrets per Section 14.6.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import UTC, datetime
from enum import Enum
from typing import Any


class ProvenanceSource(str, Enum):
    LOCAL = "local"
    NIM = "nim"
    NIM_PARTIAL_FALLBACK = "nim_partial_fallback"


@dataclass
class ProvenanceModel:
    """
    Canonical provenance object for UI requests and responses per Section 14.6.
    Contains no secrets or sensitive authentication data.
    """

    source: ProvenanceSource
    provider: str = "nvidia_nim"
    model_id: str | None = None
    base_url_label: str | None = None
    generated_at: str = field(default_factory=lambda: datetime.now(UTC).isoformat())
    request_id: int | None = None
    source_revision: str | None = None
    fallback_scene_ids: list[int] = field(default_factory=list)
    validation_warnings: list[str] = field(default_factory=list)

    def to_dict(self) -> dict[str, Any]:
        """Convert provenance object to dict without secrets."""
        return {
            "source": (
                self.source.value if isinstance(self.source, ProvenanceSource) else str(self.source)
            ),
            "provider": self.provider,
            "model_id": self.model_id,
            "base_url_label": self.base_url_label,
            "generated_at": self.generated_at,
            "request_id": self.request_id,
            "source_revision": self.source_revision,
            "fallback_scene_ids": list(self.fallback_scene_ids),
            "validation_warnings": list(self.validation_warnings),
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> ProvenanceModel:
        """Create ProvenanceModel from a dictionary."""
        source_val = data.get("source", ProvenanceSource.LOCAL)
        if isinstance(source_val, str):
            try:
                source_val = ProvenanceSource(source_val)
            except ValueError:
                source_val = ProvenanceSource.LOCAL

        return cls(
            source=source_val,
            provider=data.get("provider", "nvidia_nim"),
            model_id=data.get("model_id"),
            base_url_label=data.get("base_url_label"),
            generated_at=data.get("generated_at") or datetime.now(UTC).isoformat(),
            request_id=data.get("request_id"),
            source_revision=data.get("source_revision"),
            fallback_scene_ids=data.get("fallback_scene_ids", []),
            validation_warnings=data.get("validation_warnings", []),
        )

    @classmethod
    def create_local(
        cls, request_id: int | None = None, source_revision: str | None = None
    ) -> ProvenanceModel:
        """Create a default local deterministic provenance record."""
        return cls(
            source=ProvenanceSource.LOCAL,
            provider="local_deterministic",
            request_id=request_id,
            source_revision=source_revision,
        )
